app/controllers/training_data/code_tags_samples_controller.rb

```diff
@@ -0,0 +1,38 @@
+class TrainingData::CodeTagsSamplesController < ApplicationController
+  before_action :use_track, only: [:next]
+  before_action :use_sample, only: [:show]
+  before_action :ensure_trainer!
+
+  def index; end
+
+  def show; end
+
+  def next
+    sample = TrainingData::CodeTagsSample::RetrieveNext.(@track, params[:status])
+    return redirect_to training_data_root_path if sample.nil?
+
+    sample.lock_for_editing!(current_user)
+
+    redirect_to training_data_code_tags_sample_path(sample, status: params[:status])
+  end
+
+  private
+  def use_track
+    @track = Track.for!(params[:track])
+  rescue ActiveRecord::RecordNotFound
+    render_404(:track_not_found)
+  end
+
+  def use_sample
+    @sample = TrainingData::CodeTagsSample.find_by!(uuid: params[:id])
+    @track = @sample.track
+  rescue ActiveRecord::RecordNotFound
+    render_404(:sample_not_found)
+  end
+
+  def ensure_trainer!
+    return if current_user.trainer?(@track)
+
+    render_403(:not_trainer)
+  end
```

Comment:
This isn't an API controller, so these `render_xxx` things aren't valid. They need to redirect to actual UI pages.

Reply:
The next controller has the correct one for `ensure_trainer` btw.

Reply:
`render_404` is defined in application controller but doesn't take an argument (it just renders a 404 page). `render_403` is not defined there at all.